In [2]:
#importing the dependencies
import pandas as pd
import numpy as np
import seaborn as sns

In [3]:
data = pd.read_csv("creditcard.csv")
data.shape

(284807, 31)

In [3]:
from sklearn.model_selection import train_test_split

In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [6]:
#data information
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [7]:
#cheaking the number of missing values in each columns
data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [9]:
# the distribution of legit trasaction
data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

ABOVE DATA SET IS VERY UNBALENCED I.E more than 95% of the data on non fraudulent and 5% of the data is fraudulent

In [10]:
#saparating the data for analysis
legit_trans = data[data.Class == 0]
fraud_trans = data[data.Class == 1]

In [16]:
#legit_trans['Class'].value_counts()
legit_trans.shape

(284315, 31)

In [17]:
#statistical measures of the legit transaction
legit_trans.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

25% means 25% of values below  5.650000
and same in the case of 50% and 75%

In [18]:
#statistical measures of the fraud transaction
fraud_trans.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [19]:
#compare the values for both transactions
data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

In [20]:
#for unbalenced data we have to do under sampling


build the sample data set by  containing similar distribution legit and fraudulent transaction

In [21]:
# number of fraulent transactions is 492 and we will take random 492 transaction from legit transaction

In [22]:
legit_sample = legit_trans.sample(n = 492)

In [25]:
legit_sample['Class'].value_counts()

0    492
Name: Class, dtype: int64

In [32]:
#concatinate 2 data frames
new_data = pd.concat([legit_sample,fraud_trans],axis = 0)



if we mention axis = 0 then all the values added below legit_sample and if we mention axis = 1 values added in column wise, we dont need that. axis = 0 means rows and axis = 1 means columns

In [33]:
new_data['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [34]:
new_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      93695.861789 -0.092380  0.248628  0.062385  0.080572  0.004043   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.017928  0.074064 -0.017146  0.004845  ...  0.002048  0.050949   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.015505  0.018068 -0.044505  0.014105  0.027033  0.004728  0.013604   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       78.941931  
1      122.211321  

[2 rows x 30 columns]

In [35]:
#spliting the data into features and labels
x = new_data.drop(columns = 'Class', axis = 1)
y = new_data['Class']

In [41]:
#to split the data into training data and test data
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.2,stratify = y, random_state = 2)

stratify  = y used to distribute evenly in x_train,x_test

In [45]:
x_test.shape

(197, 30)

In [48]:
#model training 
model = LogisticRegression()

In [49]:
#training logistic regression model with training data
model.fit(x_train,y_train)

C:\Users\NAVEEN KUMAR\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [50]:
#model evaluation
#accurecy on training data 
x_train_prediction = model.predict(x_train)
training_data_accurecy  = accuracy_score(x_train_prediction,y_train)

x_train_prediction is the predicting the labels with train data 
training_data_accurecy is compareing labels with y_train labels that gives accurecy

In [54]:
print("Accurecy on traing data: ",training_data_accurecy)

Accurecy on traing data:  0.9428208386277002


In [52]:
#accurecy on test data
x_test_prediction = model.predict(x_test)
test_data_accurecy = accuracy_score(x_test_prediction,y_test)

In [55]:
print("Accurecy on test data: ",test_data_accurecy)

Accurecy on test data:  0.9137055837563451


if accurecy score of trainig data is very different from test data then model is OVER FITTED OR UNDER FITTED

IF accuecy score training data >  test data then model is over fitted
if accuracy score of training data < test data than model is under fitted

we dont want both to happen